### ISSUE TITLE
VictorLogs Bugs: Read timeout with stream_context before or after 

### ISSUE THREAD

**Issue by worker24h** [VictoriaMetrics/VictoriaMetrics#7766]
### Describe the bug

![image](https://github.com/user-attachments/assets/faf2ddeb-9ec0-49d1-8a40-42717ce0a79b)

Reference issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7637
Optimize  PR: https://github.com/VictoriaMetrics/VictoriaMetrics/commit/bddb0e369f7b815046c3935d3aee6134bb35f889

@valyala @jiekun  Hello, the case doesn't improve with the latest version in my environment.   

### To Reproduce

In 800 millions of millions of daily logs

### Version

version: v1.3.2-victorialogs

**Comment by valyala** [comment]
@jiekun , could you take a look at this issue?

**Comment by jiekun** [comment]
Sure! @worker24h  May I ask how many logs you are expecting to query by this LogsQL: `MarsUploadServiceImpl`? I think the optimization does work but if you still have too many logs in the filter result * and related stream, it could be slow. The complexity of the match operation could be `O(row count of filter result * row count of related stream)`.

**Comment by valyala** [comment]
@worker24h , could you provide the results of the following query on the time range when you see timeout for the [`stream_context` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#stream_context-pipe)?

```logsql
_time:5m MarsUploadServiceImpl | count() rows, count_uniq(_stream_id) streams
```

**Comment by worker24h** [comment]
> _time:5m MarsUploadServiceImpl | count() rows, count_uniq(_stream_id) streams

As follow 
![image](https://github.com/user-attachments/assets/853a49ac-31cd-4b16-b1eb-4b019bb18d43)

@valyala @jiekun

**Comment by worker24h** [comment]
> > _time:5m MarsUploadServiceImpl | count() rows, count_uniq(_stream_id) streams
> 
> As follow ![image](https://private-user-images.githubusercontent.com/24838476/393861386-853a49ac-31cd-4b16-b1eb-4b019bb18d43.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NTMwODgsIm5iZiI6MTczMzc1Mjc4OCwicGF0aCI6Ii8yNDgzODQ3Ni8zOTM4NjEzODYtODUzYTQ5YWMtMzFjZC00YjE2LWIxZWItNGIwMTliYjE4ZDQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDEzNTk0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2NTJhNzUwNjkzZjViZWY4ZDEwY2ZlZmFiNjUxMmMwZjUyY2ZhY2FjMTliYTMwOTAzZWEwMmUxYWVkYjIxY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X6F5bTHWQu3bUWfBXgDIcDZBwVsjDkvjPq9SlTfFnI4)
> 
> @valyala @jiekun

In our cluster, 800 million logs are generated every day

**Comment by jiekun** [comment]
Hello @worker24h sorry for the late reply. 

After discussions, we would like to provide some explanations regarding the `stream_context` method. Initially, we expect users using it in such scenarios:
1. Firstly, finding a small number of logs within a short time range (e.g., < 10 entries) through search.
2. For these few logs, users would want to know their before/after log entries.

As a result, the current implementation of this method is as follows:
1. Retrieve a small number of logs (let's call them `rows1`).
2. Fetch logs with the same stream_id as those in rows1, within a +/- 1h time range (`rows2`).
3. For each log in `rows1`, iterate through `rows2` to find its `before`/`after` log entries.

This approach leads to a significant increase in execution time if the lengh of `rows1` and `rows2` are high, due to the nested execution of the `for` loop for searching.

Therefore, our future optimization direction will be:
1. continue to guide users to utilize `stream_context` for only a small number of logs. For example, returns an error when using `stream_context` for more than xx logs/xx streams.
2. optimizing the query algorithm of stream_context (as a secondary focus).

I plan to finish the first one on Friday. Any feedback and implementation suggestions are also welcome.

**Comment by worker24h** [comment]
Thanks，your answer.

![image](https://github.com/user-attachments/assets/1ccfb2ec-44c7-4faa-86db-1d1f93ae4317)
If I have assigned the ` _stream_id ` in LogQL query, but the response is still timeout. 
Here, I have some suggestions or ideas, may be not correct：
- Firstly, the log context is very important, We have to improve it so that it can be used product online
- The root cause is too many entry logs, so I think we can reduce data from disk at the first phase，e.g. `in scanning phase`
   - We can push down the predicate `_time` or `limit` before other operators，so there's going to be a  few data 
   - Refactor this part of the code logic and data structure，for example，used `map` to replace `nested for`
- The first optimization which the error toast is very nice, but it does't set too small, because the traffic is very large in China

I'll be happy to help if you need it

@jiekun @valyala

**Comment by valyala** [comment]
The read timeout is resolved in the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/afd926a0b0d52608fe51d754aab00c14b98eeb0a by limiting the number of logs and [log streams](https://docs.victoriametrics.com/victorialogs/keyconcepts/#stream-fields), which can be passed to [`stream_contex` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#stream_context-pipe). See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7903#issuecomment-2558464858 for details.

This commit has been included in [v1.4.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/), so closing this issue as addressed.
